### PR TITLE
Static typing for profilers

### PR DESCRIPTION
**Pull request by tonywu315**
#608

### CODE REVIEW

File: dataprofiler/profilers/base_column_profilers.py [hunk 1]
```diff
@@ -36,12 +36,12 @@ def __init__(self, name: Optional[str]) -> None:
         :param name: Name of the dataset
         :type name: String
         """
-        self.name = name
+        self.name: Optional[str] = name
```
Comment (JGSweets) [inline dataprofiler/profilers/base_column_profilers.py:39]: I'm not sure this is allowed to be optional. Reason for choosing `Optional` here?

File: dataprofiler/profilers/base_column_profilers.py [hunk 2]
```diff
@@ -251,7 +251,9 @@ def report(self, remove_disabled_flag: bool = False) -> None:
         raise NotImplementedError()
 
 
-class BaseColumnPrimitiveTypeProfiler(with_metaclass(abc.ABCMeta, BaseColumnProfiler)):
+class BaseColumnPrimitiveTypeProfiler(
+    with_metaclass(abc.ABCMeta, BaseColumnProfiler)  # type: ignore
+):
     """Abstract class for profiling primative data type for col of data."""
 
     def __init__(self, name: Optional[str]) -> None:
```
Comment (JGSweets) [inline dataprofiler/profilers/base_column_profilers.py:259]: I'm not sure `name` is allowed to be optional. Reason for choosing Optional here?

File: dataprofiler/profilers/categorical_column_profile.py [hunk 3]
```diff
@@ -22,7 +27,7 @@ class CategoricalColumn(BaseColumnProfiler):
     # Default value that determines if a given col is categorical or not.
     _CATEGORICAL_THRESHOLD_DEFAULT = 0.2
 
-    def __init__(self, name, options=None):
+    def __init__(self, name: Optional[str], options: CategoricalOptions = None) -> None:
```
Comment (JGSweets) [inline dataprofiler/profilers/categorical_column_profile.py:30]: I'm not sure `name` is allowed to be optional. Reason for choosing Optional here?

File: dataprofiler/profilers/base_column_profilers.py (hunk 1 above)
Comment (JGSweets) [reply dataprofiler/profilers/base_column_profilers.py:39]: nvm it is the tests, right?

File: dataprofiler/profilers/base_column_profilers.py (hunk 2 above)
Comment (JGSweets) [reply dataprofiler/profilers/base_column_profilers.py:259]: ahh do the tests do this?

File: dataprofiler/profilers/categorical_column_profile.py (hunk 3 above)
Comment (JGSweets) [reply dataprofiler/profilers/categorical_column_profile.py:30]: as above

File: dataprofiler/profilers/float_column_profile.py [hunk 4]
```diff
@@ -285,7 +285,7 @@ def _get_float_precision(
         return subset_precision
 
     @classmethod
-    def _is_each_row_float(cls, df_series: pd.Series) -> List[bool]:
+    def _is_each_row_float(cls, df_series: pd.Series[float]) -> List[bool]:
```
Comment (JGSweets) [inline dataprofiler/profilers/float_column_profile.py:288]: I think pd.Series actually has to be an object or str
`pd.Series[str]` or `pd.Series[Object]`

Which is the intent of `NumericStatsMixin.is_float` to determine fro that what a is numeric

File: dataprofiler/profilers/graph_profiler.py [hunk 5]
```diff
@@ -220,60 +225,88 @@ def update(self, graph: nx.Graph) -> GraphProfiler:
...
-    ):
+        self,
+        graph: nx.Graph,
+        prev_dependent_properties: Dict = None,
+        subset_properties: Dict = None,
+    ) -> None:
         """Update avg_node_degree for profile."""
         self._avg_node_degree = self._get_avg_node_degree(
-            graph, subset_properties["num_nodes"]
+            graph, subset_properties["num_nodes"]  # type: ignore
```
Comment (JGSweets) [inline dataprofiler/profilers/graph_profiler.py:254]: why do we need `# type: ignore` here?

File: dataprofiler/profilers/profile_builder.py [hunk 6]
```diff
@@ -952,7 +951,7 @@ def _remove_data_labelers(
             if profiler and "data_labeler" in profiler._profiles:
                 profiler._profiles["data_labeler"].data_labeler = replacement_type
 
-        return data_labeler
+        return cast(Optional[DataLabeler], data_labeler)
```
[the remark this answers is not in the document]
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:954]: Can we do this w/o a cast?

**Review by JGSweets — CHANGES_REQUESTED** [review]
Few changes requested

File: dataprofiler/profilers/base_column_profilers.py (hunk 1 above)
Comment (tonywu315) [reply dataprofiler/profilers/base_column_profilers.py:39]: Some of the subclasses initialize the profiler with `None`. For example, `categorical_column_profile.py:66` is `merged_profile = CategoricalColumn(None)`.

File: dataprofiler/profilers/categorical_column_profile.py (hunk 3 above)
Comment (tonywu315) [reply dataprofiler/profilers/categorical_column_profile.py:30]: same reason as above in the `__add__` method

File: dataprofiler/profilers/data_labeler_column_profile.py [hunk 7]
```diff
@@ -336,9 +336,11 @@ def diff(self, other_profile: DataLabelerColumn, options: Dict = None) -> Dict:
 
         differences = {
             "data_label": utils.find_diff_of_lists_and_sets(labels, other_labels),
-            "avg_predictions": utils.find_diff_of_dicts(avg_preds, other_avg_preds),
+            "avg_predictions": utils.find_diff_of_dicts(
+                avg_preds, other_avg_preds  # type: ignore
```
[the remark this answers is not in the document]
Comment (tonywu315) [reply dataprofiler/profilers/data_labeler_column_profile.py:340]: `avg_preds` and `other_avg_preds` can both be `None` and `utils.find_diff_of_dicts` does not take in `None`

File: dataprofiler/profilers/graph_profiler.py (hunk 5 above)
Comment (tonywu315) [reply dataprofiler/profilers/graph_profiler.py:254]: `subset_properties`'s default value is `None` which is not subscriptable

File: dataprofiler/profilers/numerical_column_stats.py [hunk 8]
```diff
@@ -9,9 +9,9 @@
 from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 
 import numpy as np
-import pandas as pd
 import scipy.stats
 from future.utils import with_metaclass
+from pandas import Series
```
[the remark this answers is not in the document]
Comment (tonywu315) [reply dataprofiler/profilers/numerical_column_stats.py:14]: `pandas` wasnt used for anything else. I'll change it back to be consistent.

File: dataprofiler/profilers/profile_builder.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tonywu315) [reply dataprofiler/profilers/profile_builder.py:954]: `data_labeler` has type `BaseDataLabeler`, but `_remove_data_labelers` and several other functions expect a return of `DataLabeler`.

File: dataprofiler/profilers/float_column_profile.py (hunk 4 above)
Comment (tonywu315) [reply dataprofiler/profilers/float_column_profile.py]: oh I see, I reverted it

File: dataprofiler/profilers/base_column_profilers.py (hunk 1 above)
Comment (JGSweets) [reply dataprofiler/profilers/base_column_profilers.py:39]: Yup, makes sense.

File: dataprofiler/profilers/data_labeler_column_profile.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (JGSweets) [reply dataprofiler/profilers/data_labeler_column_profile.py:340]: Would this not indicate a code bug if `find_diff_of_dicts` cannot handle `None`, but we might be inserting it?

File: dataprofiler/profilers/graph_profiler.py (hunk 5 above)
Comment (JGSweets) [reply dataprofiler/profilers/graph_profiler.py:254]: Is there ever a chance where that could get passed to here?

File: dataprofiler/profilers/profile_builder.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:954]: Should we instead allow the abstract version BaseDataLabeler?

File: dataprofiler/profilers/data_labeler_column_profile.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tonywu315) [reply dataprofiler/profilers/data_labeler_column_profile.py:340]: I changed `find_diff_of_dicts` to handle `None` in the same way that the other `find_diff` functions do.

File: dataprofiler/profilers/graph_profiler.py (hunk 5 above)
Comment (tonywu315) [reply dataprofiler/profilers/graph_profiler.py:254]: I think `{}` is usually passed in for `subset_properties`, so I changed the default value from `None` to `{}` and removed all of the `# type: ignore`.

File: dataprofiler/profilers/profile_builder.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tonywu315) [reply dataprofiler/profilers/profile_builder.py:954]: I think thats a good idea.

**Review by JGSweets — CHANGES_REQUESTED** [review]

**Review by JGSweets — APPROVED** [review]

**Review by taylorfturner — APPROVED** [review]

**Pull request merged**